src/gui/models/isotopologueSetModel.cpp

```diff
@@ -13,6 +13,104 @@ void IsotopologueSetModel::setSourceData(IsotopologueSet &set)
     endResetModel();
 }
 
+std::vector<QModelIndex> IsotopologueSetModel::addMissingSpecies(const std::vector<std::unique_ptr<Species>> &species)
+{
+    if (!set_)
+        return {};
+
+    auto &set = set_->get();
+
+    auto nAdded = 0;
+    for (const auto &sp : species)
+    {
+        if (!set.contains(sp.get()))
+        {
+            beginInsertRows({}, set.nSpecies(), set.nSpecies());
+            set.add(sp->naturalIsotopologue(), 1.0);
+            endInsertRows();
+            ++nAdded;
+        }
+    }
+
+    if (nAdded > 0)
+    {
+        emit(dataChanged(createIndex(set.nSpecies() - nAdded, 0), createIndex(set.nSpecies() - 1, 0)));
+        std::vector<QModelIndex> newIndices;
+        for (auto n = set.nSpecies() - nAdded; n < set.nSpecies(); ++n)
+            newIndices.push_back(createIndex(n, 0));
+        return newIndices;
+    }
+
+    return {};
```

Comment:
```suggestion
    std::vector<QModelIndex> newIndices;
    for (auto n = set.nSpecies() - nAdded; n < set.nSpecies(); ++n)
        newIndices.push_back(createIndex(n, 0));
    if (!newIndices.empty())
        emit(dataChanged(createIndex(set.nSpecies() - nAdded, 0), createIndex(set.nSpecies() - 1, 0)));
    return newIndices;
```

This arrangement does reduce the complexity slightly.